README.md
# ATZAlertController

`ATZAlertController` helps you to create an alert from any class without any dependency on `ViewController`. The name itself derives that it provides A to Z solution towards the `UIAlertController` with various types of functions you can use them accordingly. `ATZAlertController` is easy to use and implement.  

# Pod Installation 

CocoaPods is the preferred way to install this library. Add this command to your `Podfile`:

##### For Objective C 
***
```
pod 'ATZAlertController-ObjC'
```

###[For Swift](https://github.com/shashankpali/ATZAlertController).

# Direct Installation

Download .zip file and extract it. There you will find the `ATZAlertController` folder, copy all the files from the folder to your respective project.

# Integration Method
***
```objc
#import "ATZAlertController.h"

[ATZAlertController alertWithTitle:@"My Alert"
                           message:@"Array containing object string"
                 cancelButtonTitle:@"Cancel"
            destructiveButtonTitle:@"Remove"
                 otherButtonTitles:@[@"Ok", @"Right"]
                    preferredStyle:UIAlertControllerStyleAlert
                showAlertInstantly:YES
                             block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString) {
                                //your code
                              }];
```

```objc
#import "ATZAlertController.h"

ATZActionButton *button1 = [ATZActionButton actionWithTitle:@"Ok" preferredStyle:UIAlertActionStyleDefault];
ATZActionButton *button2 = [ATZActionButton actionWithTitle:@"Remove" preferredStyle:UIAlertActionStyleDestructive];

ATZAlertController *alert = [ATZAlertController alertWithTitle:@"My Alert"
                                   message:@"Array containing object string"
                         cancelButtonTitle:@"Cancel"
                         otherActionButton:@[button1, button2]
                            preferredStyle:UIAlertControllerStyleAlert
                        showAlertInstantly:NO
                                     block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString) {
                                        //your code
                                      }];

[alert showWithAnimation:NO];
```

### Thanks
- Reference from [ASJAlertController](https://github.com/sudeepjaiswal/ASJAlertController).

# License

`ATZAlertController` is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

## [1.1](https://github.com/shashankpali/ATZAlertController/releases/tag/1.1)
Released on Tuesday, 2 August, 2016.

#### Updated
* Updated doumentation.

#### Fixed
* Fixed unused variable warnings.

## [1.0](https://github.com/shashankpali/ATZAlertController/releases/tag/1.0)
Released on Saturday, 30 July, 2016.

#### Added
* Initial release.

ATZAlertController/ATZAlertController.m
// ATZAlertController.m
//
// Copyright (c) 2016 Shashank Pali
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "ATZAlertController+Helper.h"

@implementation ATZAlertController

- (void)show
{
  [self showAlertAnimation:YES];
}

- (void)showWithAnimation:(BOOL)animated
{
  [self showAlertAnimation:animated];
}

#pragma mark - View Disapper

- (void)viewDidDisappear:(BOOL)animated
{
  [super viewDidDisappear:animated];
  
  // precaution to insure window gets destroyed
  self.alertWindow.hidden = YES;
  self.alertWindow = nil;
}

#pragma mark - instance Method

+ (nonnull instancetype)alertWithTitle:(NSString *)title
                               message:(NSString *)message
                     cancelButtonTitle:(NSString *)cancel
                destructiveButtonTitle:(NSString *)destructive
                     otherButtonTitles:(NSArray<NSString *> *)titleArray
                        preferredStyle:(UIAlertControllerStyle)style
                    showAlertInstantly:(BOOL)showAlert
                                 block:(actionHandler)handler
{
  for (id __attribute__((unused)) obj in titleArray) {
    
    NSAssert([obj isKindOfClass:[NSString class]], @"Required array containing \"String\" as object");
  }
  
  ATZAlertController *alert = [ATZAlertController title:title message:message cancelButtonTitle:cancel destructiveButtonTitle:destructive otherButtonTitles:titleArray preferredStyle:style block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString)
                               {
                                 if (handler) {
                                   handler(action , titleString);
                                 }
                               }];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

+ (nonnull instancetype)alertWithTitle:(NSString * _Nonnull)title
                               message:(NSString * _Nonnull)message
                     cancelButtonTitle:(NSString * _Nonnull)cancel
                     otherActionButton:(NSArray<ATZActionButton *> * _Nonnull)buttonArray
                        preferredStyle:(UIAlertControllerStyle)style
                    showAlertInstantly:(BOOL)showAlert
                                 block:(nullable actionHandler)handler
{
  for (id __attribute__((unused)) obj in buttonArray) {
    
    NSAssert([obj isKindOfClass:[ATZActionButton class]], @"Required array containing \"ATZActionButton\" as object");
  }
  
  ATZAlertController *alert =[ATZAlertController title:title message:message cancelButtonTitle:cancel destructiveButtonTitle:nil otherButtonTitles:buttonArray preferredStyle:style block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString)
                              {
                                if (handler) {
                                  handler(action , titleString);
                                }
                              }];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

+ (nonnull instancetype)alertWithTitle:(NSString * _Nonnull)title
                               message:(NSString * _Nonnull)message
                     cancelButtonTitle:(NSString * _Nonnull)cancel
                    showAlertInstantly:(BOOL)showAlert
{
  ATZAlertController *alert =[ATZAlertController title:title message:message cancelButtonTitle:cancel destructiveButtonTitle:nil otherButtonTitles:nil preferredStyle:UIAlertControllerStyleAlert block:nil];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

+ (nonnull instancetype)loginAlertWithTitle:(NSString *)title
                                    message:(NSString *)message
                   userNamePlaceholderTitle:(NSString *)usernameTitle
                   passwordPlaceholderTitle:(NSString * _Nonnull)passwordTitle
                         showAlertInstantly:(BOOL)showAlert
                                      block:(nullable actionHandler)handler
{
  ATZAlertController *alert = [ATZAlertController title:title message:message cancelButtonTitle:nil destructiveButtonTitle:nil otherButtonTitles:@[@"Cancel",@"Login"] preferredStyle:UIAlertControllerStyleAlert block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString) {
    
    
  }];
  [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
    textField.placeholder = usernameTitle;
  }];
  [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
    textField.placeholder = passwordTitle;
    textField.secureTextEntry = YES;
  }];
  if (showAlert)
  {
    [alert show];
  }
  return alert;
}

@end

#pragma mark - ATZActionButton

@implementation ATZActionButton

+ (ATZActionButton * _Nonnull)actionWithTitle:(NSString * _Nonnull)title preferredStyle:(UIAlertActionStyle)style{
  
  ATZActionButton *alertButton = [[ATZActionButton alloc] init];
  alertButton.title = title;
  alertButton.style = style;
  
  return alertButton;
}

@end

Example/ATZAlertController-ObjC/AppDelegate.m
//
//  AppDelegate.m
//  ATZAlertController-ObjC
//
//  Created by Shashank Pali on 30/07/16.
//  Copyright © 2016 Shashank Pali. All rights reserved.
//

#import "AppDelegate.h"

@interface AppDelegate ()

@end

@implementation AppDelegate


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  // Override point for customization after application launch.
  return YES;
}

- (void)applicationWillResignActive:(UIApplication *)application {
  // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
  // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
  // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
  // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application {
  // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
  // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application {
  // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Example/ATZAlertController-ObjC/ViewController.m
//
//  ViewController.m
//  ATZAlertController
//
//  Created by shashank on 30/07/16.
//  Copyright © 2016 shashank. All rights reserved.
//

#import "ViewController.h"
#import "ATZAlertController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];
  // Do any additional setup after loading the view, typically from a nib.
  
}

- (IBAction)testBtn:(id)sender {
  
  UIButton *button = (UIButton *)sender;
  ATZAlertController *alert= nil;
  switch (button.tag) {
    case 0:
      alert = [ATZAlertController alertWithTitle:@"Simple Alert" message:@"show my alert" cancelButtonTitle:@"ok" showAlertInstantly:NO];
      
      [alert show];
      
      break;
    case 1:
      [ATZAlertController alertWithTitle:@"My Alert"
                                 message:@"Array containing object string"
                       cancelButtonTitle:@"Cancel"
                  destructiveButtonTitle:@"Remove"
                       otherButtonTitles:@[@"Ok",@"Right"]
                          preferredStyle:UIAlertControllerStyleAlert
                      showAlertInstantly:YES
                                   block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString) {
                                     
                                     if ([titleString isEqualToString:@"Cancel"]) {
                                       
                                       NSLog(@"Cancel");
                                     }
                                     if ([titleString isEqualToString:@"Remove"]) {
                                       
                                       NSLog(@"Remove");
                                     }
                                     if ([titleString isEqualToString:@"Ok"]) {
                                       
                                       NSLog(@"Ok");
                                     }
                                     if ([titleString isEqualToString:@"Right"]) {
                                       
                                       NSLog(@"Right");
                                     }
                                     
                                   }];
      
      break;
    case 2: {
      ATZActionButton *button1 = [ATZActionButton actionWithTitle:@"Ok" preferredStyle:UIAlertActionStyleDefault];
      ATZActionButton *button2 = [ATZActionButton actionWithTitle:@"Remove" preferredStyle:UIAlertActionStyleDestructive];
      
      alert = [ATZAlertController alertWithTitle:@"My Alert"
                                         message:@"Array containing object ATZActionButton"
                               cancelButtonTitle:@"Cancel"
                               otherActionButton:@[button1,button2]
                                  preferredStyle:UIAlertControllerStyleAlert
                              showAlertInstantly:NO
                                           block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString) {
                                             
                                             if ([titleString isEqualToString:@"Cancel"]) {
                                               
                                               NSLog(@"Cancel");
                                             }
                                             if ([titleString isEqualToString:@"Remove"]) {
                                               
                                               NSLog(@"Remove");
                                             }
                                             if ([titleString isEqualToString:@"Ok"]) {
                                               
                                               NSLog(@"Ok");
                                             }
                                           }];
      
      [alert showWithAnimation:NO];
      
      break;
    }
    case 3:{
      
      [ATZAlertController loginAlertWithTitle:@"Login" message:nil userNamePlaceholderTitle:@"Username" passwordPlaceholderTitle:@"Password" showAlertInstantly:YES block:^(UIAlertAction * _Nonnull action, NSString * _Nonnull titleString) {
        
        
      }];
      
      break;
    }
      
    default:
      break;
  }
  
}


- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
  // Dispose of any resources that can be recreated.
}

@end
